File: AtCoder/ABC151/D.py
H, W = map(int, input().split())
M = [[-1]*(W+2) for _ in range(H+2)]
X = []
for h in range(H):
    for w, c in enumerate(input()):
        if c=='.':
            M[h+1][w+1] = 0
            X.append((h+1, w+1))

from pprint import pprint
pprint(M)

mark = 0
ans = 0
for ist, st in enumerate(X):
    for gl in X[ist+1:]:
        mark += 1
        turn = 0
        f = False
        queue = [st]
        while queue and not f:
            queue_new = []
            for qx, qy in queue:
                if (qx, qy)==gl:
                    f = True
                    break
                    
                for dx, dy in ((+1,0),(-1,0),(0,+1),(0,-1)):
                    x = qx+dx
                    y = qy+dy
                    m = M[x][y] 
                    if m!=-1 and m!=mark:
                        M[x][y] = mark
                        queue_new.append((x,y))
                        
            
            queue = queue_new
            turn += 1

        ans = max(ans, turn-1)

print(ans)

File: AtCoder/ABC151/D.cpp
#include<iostream>
#include<vector>
using namespace std;
#define ll long long

int main(){
  ll H,W;
  cin >> H >> W;
  vector<ll> row(W+2, -1);
  vector<vector<ll>> M(H+2, row);
  vector<vector<ll>> X;
  
  for(ll h=0; h<H; ++h) {
    string s;
    cin >> s;
    for(ll w=0; w<W; ++w) {
      if(s[w]=='#') continue;
      M[h+1][w+1] = 0;
      vector<ll> x = {h+1, w+1};
      X.push_back(x);
    }
  }

  //for(ll h=0; h<H+2; ++h) {
  //for(ll w=0; w<W+2; ++w)
  //  cout << -M[h][w]; cout << endl;}


  ll mark = 0, ans = 0;
  for(ll ist=0    ; ist<X.size(); ++ist)  { 
    ++mark;
    ll turn = 0;
    bool f = false;
    vector<vector<ll>> queue = {X[ist]};
    M[X[ist][0]][X[ist][1]] = mark;

    while(queue.size() && !f) {
      vector<vector<ll>> queue_new;
      for(ll iq=0; iq<queue.size(); ++iq) {
        ll qx = queue[iq][0], qy = queue[iq][1];
        
        //cout << ist << ":" << turn << ":" << qx << "," <<  qy << endl;
        
        ll x, y, m;
        x = qx+1; y = qy  ; m = M[x][y];
        if(m!=-1 && m!=mark) { M[x][y] = mark; vector<ll> xx = {x,y}; queue_new.push_back(xx); }
        x = qx-1; y = qy  ; m = M[x][y];
        if(m!=-1 && m!=mark) { M[x][y] = mark; vector<ll> xx = {x,y}; queue_new.push_back(xx); }
        x = qx  ; y = qy+1; m = M[x][y];
        if(m!=-1 && m!=mark) { M[x][y] = mark; vector<ll> xx = {x,y}; queue_new.push_back(xx); }
        x = qx  ; y = qy-1; m = M[x][y];
        if(m!=-1 && m!=mark) { M[x][y] = mark; vector<ll> xx = {x,y}; queue_new.push_back(xx); }
        
      }

      queue = queue_new;
      ++turn;
    }

    ans = max(ans, turn-1);

  }

  cout << ans << endl;

  return 0;
}
